core/src/main/scala/org/apache/spark/rdd/UnionRDD.scala

```diff
@@ -62,7 +64,23 @@ class UnionRDD[T: ClassTag](
     var rdds: Seq[RDD[T]])
   extends RDD[T](sc, Nil) {  // Nil since we implement getDependencies
 
+  // Evaluate partitions in parallel. Partitions of each rdd will be cached by the `partitions`
+  // val in `RDD`.
+  private[spark] lazy val parallelPartitionEval: Boolean = {
```

Comment:
I think there's a big difference between the two scenarios. While we may have lurking concurrency bugs in getSplits calls today, this patch kicks off a bunch of tasks that call getSplits at the same time. That's much more likely to hit concurrency problems if they exist and I don't think we can conclude that it is safe from your example, where a driver could be doing split calculations in multiple threads.